core/src/subgraph/instance.rs

```diff
@@ -82,17 +99,49 @@ where
 
         // Process the log in each host in the same order the corresponding
         // data sources appear in the subgraph manifest
-        Box::new(stream::iter_ok(matching_hosts).fold(
-            entity_operations,
-            move |entity_operations, host| {
+        Box::new(
+            stream::iter_ok(matching_hosts).fold(state, move |state, host| {
                 host.process_log(
                     logger.clone(),
                     block.clone(),
                     transaction.clone(),
                     log.clone(),
-                    entity_operations,
+                    state,
                 )
-            },
-        ))
+            }),
+        )
+    }
+
+    fn ethereum_log_filter(&self) -> EthereumLogFilter {
+        self.ethereum_log_filter.clone()
+    }
+
+    fn runtime_host_builder(&self) -> &T {
+        &self.host_builder
+    }
+
+    fn add_dynamic_data_sources(
+        &mut self,
+        data_sources: Vec<DataSource>,
+        runtime_hosts: Vec<Arc<T::Host>>,
+    ) -> Result<(), Error> {
+        // Protect against creating more than the allowed maximum number of data sources
+        if let Some(max_data_sources) = *MAX_DATA_SOURCES {
+            if self.hosts.len() + runtime_hosts.len() > max_data_sources {
```

Comment:
It's a 1:1 mapping right now. I agree it could maybe be more obvious but exposing "max data sources" feels more intuitive / easier to communicate to users.

Reply:
I see. I think 1:1 data source to mapping is a good model to keep going forward, so this is fine as it is.